server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -77,7 +75,12 @@ public class Task {
 
     private final Map<Long, List<ThreadResourceInfo>> resourceStats;
 
-    private final Phaser resourceTrackingThreadsBarrier = new Phaser(1);
+    private final List<TaskResourceTrackingListener> resourceTrackingListeners;
+
+    // Keeps track of the number of active resource tracking threads for this task. It is initialized to 1 to track
+    // the task's own/self thread. When this value becomes 0, all threads have been marked inactive and the resource
+    // tracking can be stopped for this task.
+    private final AtomicInteger numActiveResourceTrackingThreads = new AtomicInteger(1);
```

Comment:
Not sure if this assumption is guaranteed f.e. can the initial thread that created the task completes execution(listener isn't invoked so the request might still be in-flight) taking the count to zero, while sub-tasks are haven't been started yet. In such a case will be run the risk of pre-maturely completing task resource tracking?

If the above statement is guaranteed not to happen, can we add assertions at Task#incrementResourceTrackingThreads to verify is the task is not completed yet and the listener still pending invocation.

If assertions already exist please point me to them.

Reply:
My assumption was that threads are guaranteed to start before task unregistration, but they may stop at a later point due to race-conditions. If this holds true, then it is guaranteed that the count never drops to zero pre-maturely.

Sure, I'll add an assertion for this once you confirm my assumption.

Reply:
Yes threads are guaranteed to start before task unregistration however the thread that registers the task might be different from the one that unregisters it

Reply:
Turns out even thread starting isn't guaranteed to happen before task unregistration. Adding this assertion made tests like [RejectionActionIT](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/action/RejectionActionIT.java) fail pretty consistently.

For the above failure, I could see that the `SearchTask` was unregistered before the runnable `AbstractSearchAsyncAction` (wrapped inside a `TaskAwareRunnable`) started its execution. The search task was not marked as cancelled either.

Reply:
I'm guessing if the following sequence of events took place:
* One task and one thread started (num active threads = 2).
* Task was unregistered (num active threads = 1).
* Existing thread before stopping, spawns another thread. Since the thread count is only incremented when the runnable executes, the thread count remains 1 so far.
* Existing thread stops (num active threads = 0; the task is marked completed).
* Newly spawned thread starts executing and runs into assertion error since the task is marked completed.

Reply:
Something worth noting – resource stats are piggy-backed with the response, so any thread usage info captured after the task is unregistered may be irrelevant (at least for now). We need a better mechanism to figure out the start/stop boundary of a task given that register/unregister/thread execution happens independently.